api/server/runner_httptrigger.go

```diff
@@ -0,0 +1,145 @@
+	app, err := s.lbReadAccess.GetAppByID(ctx, appID)
+	if err != nil {
+		return err
+	}
+
+	routePath := path.Clean(p)
+	trigger, err := s.lbReadAccess.GetTriggerBySource(ctx, appID, "http", routePath)
+
+	if err != nil {
+		return err
+	}
+
+	fn, err := s.lbReadAccess.GetFnByID(ctx, trigger.FnID)
+	// gin sets this to 404 on NoRoute, so we'll just ensure it's 200 by default.
+	c.Status(200) // this doesn't write the header yet
+
+	return s.ServeHTTPTrigger(c, app, fn, trigger)
+}
+
+// TODO it would be nice if we could make this have nothing to do with the gin.Context but meh
+// TODO make async store an *http.Request? would be sexy until we have different api format...
+func (s *Server) ServeHTTPTrigger(c *gin.Context, app *models.App, fn *models.Fn, trigger *models.Trigger) error {
+	buf := bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	writer := syncResponseWriter{
+		Buffer:  buf,
+		headers: c.Writer.Header(), // copy ref
+	}
+	defer bufPool.Put(buf) // TODO need to ensure this is safe with Dispatch?
+
+	// GetCall can mod headers, assign an id, look up the route/app (cached),
+	// strip params, etc.
+	// this should happen ASAP to turn app name to app ID
+
+	// GetCall can mod headers, assign an id, look up the route/app (cached),
+	// strip params, etc.
+
+	call, err := s.agent.GetCall(
+		agent.WithWriter(&writer), // XXX (reed): order matters [for now]
+		agent.FromHttpTriggerRequest(app, fn, trigger, c.Request),
+	)
+	if err != nil {
+		return err
+	}
+	model := call.Model()
+	{ // scope this, to disallow ctx use outside of this scope. add id for handleV1ErrorResponse logger
+		ctx, _ := common.LoggerWithFields(c.Request.Context(), logrus.Fields{"id": model.ID})
+		c.Request = c.Request.WithContext(ctx)
+	}
+
+	// TODO TRIGGERWIP  not clear this makes sense here - but it works  so...
+	if model.Type == "async" {
+
+		// TODO we should push this into GetCall somehow (CallOpt maybe) or maybe agent.Queue(Call) ?
+		if c.Request.ContentLength > 0 {
+			buf.Grow(int(c.Request.ContentLength))
+		}
+		_, err := buf.ReadFrom(c.Request.Body)
+		if err != nil {
+			return models.ErrInvalidPayload
+		}
+		model.Payload = buf.String()
+
+		err = s.lbEnqueue.Enqueue(c.Request.Context(), model)
+		if err != nil {
+			return err
+		}
+
+		c.JSON(http.StatusAccepted, map[string]string{"call_id": model.ID})
+		return nil
+	}
+
+	err = s.agent.Submit(call)
+	if err != nil {
+		// NOTE if they cancel the request then it will stop the call (kind of cool),
+		// we could filter that error out here too as right now it yells a little
+		if err == models.ErrCallTimeoutServerBusy || err == models.ErrCallTimeout {
+			// TODO maneuver
+			// add this, since it means that start may not have been called [and it's relevant]
+			c.Writer.Header().Add("XXX-FXLB-WAIT", time.Now().Sub(time.Time(model.CreatedAt)).String())
+		}
+		return err
+	}
+
+	// TODO TRIGGERWIP  Not sure I want this evil magic in my life here - shoud remove
```

Comment:
i think we need to just do this content type detection as a middleware for all endpoints on the incoming request and outgoing response if we're going to do it at all. the CRUD api needs it too, it just assumes json and doesn't error on content type or anything. anyway, feel free to remove, but i think some of the examples and maybe tests rely on this behavior for better or for worse so it may not be worth it (we can hide it). it's better than defaulting to json :)